# 02-BOW

Now that we master the preprocessing, let's make our first Bag Of Words (BOW).

We will reuse our dataset of Coldplay songs to make a BOW.

As usual, the first step is to import some libraries. So import *nltk* as well as all the libraries you will need.

In [1]:
# Import NLTK and all the needed libraries
import nltk
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

Load now the dataset in *coldplay.csv* using pandas.

In [3]:
# TODO: Load the dataset in coldplay.csv

coldplay_data = pd.read_csv('coldplay.csv')


You already know this dataset, but you can check it again if you want to refresh your memory.

In [4]:
# TODO: Explore the data
coldplay_data.head()

Artist                           Song  \
0  Coldplay                 Another's Arms   
1  Coldplay                Bigger Stronger   
2  Coldplay                       Daylight   
3  Coldplay                       Everglow   
4  Coldplay  Every Teardrop Is A Waterfall   

                                                Link  \
0            /c/coldplay/anothers+arms_21079526.html   
1          /c/coldplay/bigger+stronger_20032648.html   
2                 /c/coldplay/daylight_20032625.html   
3                 /c/coldplay/everglow_21104546.html   
4  /c/coldplay/every+teardrop+is+a+waterfall_2091...   

                                              Lyrics  
0  Late night watching tv  \nUsed to be you here ...  
1  I want to be bigger stronger drive a faster ca...  
2  To my surprise, and my delight  \nI saw sunris...  
3  Oh, they say people come  \nThey say people go...  
4  I turn the music up, I got my records on  \nI ...

Now using the *CountVectorizer* of scikit-learn, make a BOW of all the lyrics of Coldplay, and print the result.

In [6]:
# TODO: Compute a BOW
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(coldplay_data['Lyrics'])
print(X)

  (0, 793)	3
  (0, 967)	3
  (0, 1683)	3
  (0, 1616)	3
  (0, 1646)	3
  (0, 1571)	9
  (0, 96)	3
  (0, 1770)	4
  (0, 673)	3
  (0, 127)	3
  (0, 892)	15
  (0, 1772)	4
  (0, 63)	18
  (0, 66)	2
  (0, 158)	4
  (0, 996)	2
  (0, 948)	2
  (0, 1704)	4
  (0, 1523)	4
  (0, 1745)	2
  (0, 895)	2
  (0, 980)	2
  (0, 44)	16
  (0, 1032)	2
  (0, 757)	3
  :	:
  (119, 1647)	2
  (119, 1601)	3
  (119, 102)	2
  (119, 988)	6
  (119, 408)	1
  (119, 1735)	1
  (119, 632)	1
  (119, 1189)	2
  (119, 1698)	1
  (119, 893)	6
  (119, 133)	1
  (119, 33)	1
  (119, 834)	1
  (119, 1661)	6
  (119, 1408)	1
  (119, 403)	6
  (119, 669)	1
  (119, 325)	3
  (119, 719)	1
  (119, 375)	1
  (119, 841)	1
  (119, 1693)	5
  (119, 519)	1
  (119, 1081)	1
  (119, 516)	1


Now that we have the BOW matrix, we would like to have a new dataframe having the BOW for each song, and as columns the corresponding words (just as we did in the lecture at the end).

So that at the end we would end up with a dataframe containing something like the following (120 raws for 120 songs, and as many columns as words):

| | ah | adventure | ... | yeah 
|---|---|---|---|---| 
| 0 | 0 | 1 | ... | 4 |
| 1 | 8 | 0 | ... | 2 |
|...|...|...|...|...|
| 119 | 5 | 0 | ... | 8 |

In [7]:
# TODO: Create a new dataframe containing the BOW outputs and the corresponding words as columns. And print it
coldplay_bow = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())
print(coldplay_bow)

     10  2000  2gether  76543  aaaaaah  aaaaah  aaaah  about  above  achin  \
0     0     0        0      0        0       0      0      0      0      0   
1     0     0        0      0        0       0      0      0      0      0   
2     0     0        0      0        0       0      0      0      0      0   
3     0     0        0      0        0       0      0      0      0      0   
4     0     0        0      0        0       0      0      0      0      0   
..   ..   ...      ...    ...      ...     ...    ...    ...    ...    ...   
115   0     0        0      0        0       0      0      1      2      0   
116   0     0        0      0        0       0      0      0      0      0   
117   0     0        1      0        0       0      0      0      0      0   
118   0     0        0      0        0       0      0      0      0      0   
119   0     0        0      0        0       0      0      0      0      0   

     ...  yellow  yes  yesterday  yet  you  young  your  yours 

Well as you see we're still having some issue, we have some tokens that are not words, like '10' or '2000'.

To get rid of that, we could use directly regular expressions within the function. Another solution would be to make preprocessing before using the function *CountVectorizer*.

For the moment, we won't pay attention to this issue. But if you are curious and have time, you can find on google how to remove those words using the *CountVectorizer*.

Now we would like to see what are the most used words by Coldplay.

In [8]:
sum_bow = coldplay_bow.sum()
sum_bow.idxmax()

'you'

So what is the most used word? Are you surprised?

Now make a sort in order to show the 10 most used.

In [9]:
# TODO: print the 10 most used word by Coldplay
top_10_words = sum_bow.nlargest(10)
print(top_10_words)

you    994
the    777
and    650
to     481
it     458
oh     334
in     318
me     314
my     288
on     285
dtype: int64


Here it is! You know the Coldplay lyrics more than the singers now!